agent/agentcontainers/api.go

```diff
@@ -1224,6 +1229,137 @@ func (api *API) getContainers() (codersdk.WorkspaceAgentListContainersResponse,
+		Message: "Devcontainer not found.",
+		Detail:  fmt.Sprintf("Could not find devcontainer with ID: %q", devcontainerID),
+	})
+}
+
+func (api *API) handleDevcontainerDelete(w http.ResponseWriter, r *http.Request) {
+	var (
+		ctx            = r.Context()
+		devcontainerID = chi.URLParam(r, "devcontainer")
+	)
+
+	if devcontainerID == "" {
+		httpapi.Write(ctx, w, http.StatusBadRequest, codersdk.Response{
+			Message: "Missing devcontainer ID",
+			Detail:  "Devcontainer ID is required to delete a devcontainer.",
+		})
+		return
+	}
+
+	api.mu.Lock()
+
+	dc, err := api.devcontainerByIDLocked(devcontainerID)
+	if err != nil {
+		api.mu.Unlock()
+		httperror.WriteResponseError(ctx, w, err)
+		return
+	}
+
+	// Check if the devcontainer is currently starting - if so, we can't delete it.
+	if dc.Status == codersdk.WorkspaceAgentDevcontainerStatusStarting {
+		api.mu.Unlock()
+		httpapi.Write(ctx, w, http.StatusConflict, codersdk.Response{
+			Message: "Devcontainer is starting",
+			Detail:  fmt.Sprintf("Devcontainer %q is currently starting and cannot be deleted.", dc.Name),
+		})
+		return
+	}
+
+	// Similarly, if already stopping, don't allow another delete.
+	if dc.Status == codersdk.WorkspaceAgentDevcontainerStatusStopping {
+		api.mu.Unlock()
+		httpapi.Write(ctx, w, http.StatusConflict, codersdk.Response{
+			Message: "Devcontainer is stopping",
+			Detail:  fmt.Sprintf("Devcontainer %q is currently stopping.", dc.Name),
+		})
+		return
+	}
+
+	dc.Status = codersdk.WorkspaceAgentDevcontainerStatusStopping
+	dc.Error = ""
+	api.knownDevcontainers[dc.WorkspaceFolder] = dc
+	api.broadcastUpdatesLocked()
+
+	// Gather the information we need before unlocking.
+	workspaceFolder := dc.WorkspaceFolder
+	dcName := dc.Name
+	var containerID string
+	if dc.Container != nil {
+		containerID = dc.Container.ID
+	}
+	proc, hasSubAgent := api.injectedSubAgentProcs[workspaceFolder]
+	var subAgentID uuid.UUID
+	if hasSubAgent && proc.agent.ID != uuid.Nil {
+		subAgentID = proc.agent.ID
+		// Stop the subagent process context to ensure it stops.
+		proc.stop()
+	}
+
+	// Unlock the mutex while we perform potentially slow operations
+	// (network calls, docker commands) to avoid blocking other operations.
+	api.mu.Unlock()
+
+	// Stop and remove the container if it exists.
+	if containerID != "" {
+		if err := api.ccli.Stop(ctx, containerID); err != nil {
+			api.logger.Error(ctx, "unable to stop container", slog.Error(err))
+
+			httpapi.Write(ctx, w, http.StatusInternalServerError, codersdk.Response{
+				Message: "An internal error occurred stopping the container",
+				Detail:  err.Error(),
+			})
+			return
+		}
+
+		if err := api.ccli.Remove(ctx, containerID); err != nil {
+			api.logger.Error(ctx, "unable to remove container", slog.Error(err))
+
+			httpapi.Write(ctx, w, http.StatusInternalServerError, codersdk.Response{
+				Message: "An internal error occurred removing the container",
+				Detail:  err.Error(),
+			})
+			return
+		}
```

Comment:
Error paths seem to be lacking state handling for the devcontainer. Should probably update `dc.Error`, perhaps others?